tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc

```diff
@@ -0,0 +1,674 @@
+                          first_invalid_index_gpu_memory,
+                          sizeof(*first_invalid_index_host.data()))
+             .ok()) {
+      return errors::Internal("Failed to copy first_invalid_index to host");
+    }
+
+    // We must wait for num_empty_rows and rows_are_not_ordered to be copied to
+    // the host, so we enqueue the remainder of the computation onto the stream
+    // asynchronously to avoid stalling execution.
+    auto async_finish_computation =
+        [this, context, kReverseIndexMapOutput, kAllIndicesValid, index_type, N,
+         rank, indices, values, default_value, dense_rows, num_empty_rows_host,
+         rows_are_not_ordered_host, first_invalid_index_host,
+         num_empty_rows_through_t, num_empty_rows_through, input_row_ends_t,
+         input_row_ends, empty_row_indicator_t, empty_row_indicator,
+         done]() -> void {
+      CHECK(done);  // Crash OK
+
+      // Ensure that within the callback, the proper GPU settings are
+      // configured.
+      auto stream = context->op_device_context()->stream();
+      ScopedActivateExecutorContext scoped_activation{stream->parent()};
+
+      int first_invalid_index = *first_invalid_index_host.data();
+      OP_REQUIRES_ASYNC(context, first_invalid_index == kAllIndicesValid,
+                        errors::InvalidArgument("indices(", first_invalid_index,
+                                                ", 0) is invalid."),
+                        done);
+
+      Tensor* output_indices_t;
+      Tindex num_empty_rows = *num_empty_rows_host.data();
+      const Tindex N_full = N + num_empty_rows;
+      TensorShape output_indices_shape({N_full, rank});
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          context->allocate_output("output_indices", output_indices_shape,
+                                   &output_indices_t),
+          done);
+      auto output_indices = output_indices_t->matrix<Tindex>();
+
+      Tensor* output_values_t;
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          context->allocate_output("output_values", TensorShape({N_full}),
+                                   &output_values_t),
+          done);
+      auto output_values = output_values_t->vec<T>();
+
+      Tindex* reverse_index_map = nullptr;
+      if (context->output_required(kReverseIndexMapOutput)) {
+        Tensor* reverse_index_map_t = nullptr;
+        OP_REQUIRES_OK_ASYNC(
+            context,
+            context->allocate_output(kReverseIndexMapOutput, TensorShape({N}),
+                                     &reverse_index_map_t),
+            done);
+        reverse_index_map = reverse_index_map_t->vec<Tindex>().data();
+      }
+
+      const GPUDevice& device = context->eigen_device<GPUDevice>();
+
+      Tindex* input_index_map = nullptr;
+      Tensor input_index_map_t;
+      int rows_are_not_ordered = *rows_are_not_ordered_host.data();
+      if (rows_are_not_ordered) {
```

Comment:
Can some of this logic be split into separate named functions?  I think that would make the flow more readable.

Reply:
I've pulled the output allocations and sorting logic out into helper methods so the lambda is a lot shorter now.